src/sha2.c

```diff
@@ -639,3 +648,63 @@ void jtr_sha512_final(void *_output, jtr_sha512_ctx *ctx)
 	}
 #endif
 }
+
+/*********************************************************************/
+/*********************************************************************/
+/*** Here is code for reverse SHA512 rounds                        ***/
+/*********************************************************************/
+/*********************************************************************/
+void sha512_reverse(uint64_t *hash)
+{
+	uint64_t a, b, c, d, e, f, g, h, s0, maj, tmp;
+
+	a = hash[0] - INIT_A;
+	b = hash[1] - INIT_B;
+	c = hash[2] - INIT_C;
+	d = hash[3] - INIT_D;
+	e = hash[4] - INIT_E;
+	f = hash[5] - INIT_F;
+	g = hash[6] - INIT_G;
+	h = hash[7] - INIT_H;
+
+	s0 = ROR64(b, 28) ^ ROR64(b, 34) ^ ROR64(b, 39);
+	maj = (b & c) ^ (b & d) ^ (c & d);
+	tmp = d;
+	d = e - (a - (s0 + maj));
+	e = f;
+	f = g;
+	g = h;
+	a = b;
+	b = c;
+	c = tmp;
+
+	s0 = ROR64(b, 28) ^ ROR64(b, 34) ^ ROR64(b, 39);
+	maj = (b & c) ^ (b & d) ^ (c & d);
+	tmp = d;
+	d = e - (a - (s0 + maj));
+	e = f;
+	f = g;
+	a = b;
+	b = c;
+	c = tmp;
+
+	s0 = ROR64(b, 28) ^ ROR64(b, 34) ^ ROR64(b, 39);
+	maj = (b & c) ^ (b & d) ^ (c & d);
+	tmp = d;
+	d = e - (a - (s0 + maj));
+	e = f;
+	a = b;
+	b = c;
+	c = tmp;
+
+	s0 = ROR64(b, 28) ^ ROR64(b, 34) ^ ROR64(b, 39);
+	maj = (b & c) ^ (b & d) ^ (c & d);
+
+	hash[0] = e - (a - (s0 + maj));
+}
+
+void sha512_unreverse(uint64_t *hash)
+{
+	fprintf(stderr, "sha512_unreverse() not implemented\n");
+	error();
+}
```

Comment:
I suppose we can drop the unreverse function completely since we aren't exporting (nor using) it anyway.

Reply:
I agree, no reason to have a function that merely says it's unimplemented.